pre-commit hook fails for no reason

- `kcd-scripts` version: `0.0.0-semantically-released` 😉 
- `node` version: `8.5.0`
- `npm` version: `5.3.0` 
- `yarn`version: `1.0.2`

What you did:
`git commit`


What happened:

```
husky > npm run -s precommit (node v8.5.0)

25l[09:24:02] Running tasks for **/*.+(js|json|less|css|ts) [started]
[09:24:02] Running tasks for .all-contributorsrc [started]
[09:24:02] Running tasks for .all-contributorsrc [skipped]
[09:24:02] → No staged files match .all-contributorsrc
[09:24:02] kcd-scripts format [started]
[09:24:03] kcd-scripts format [completed]
[09:24:03] kcd-scripts lint [started]
[09:24:05] kcd-scripts lint [completed]
[09:24:05] kcd-scripts test --findRelatedTests [started]
[09:24:07] kcd-scripts test --findRelatedTests [completed]
[09:24:07] git add [started]
[09:24:07] git add [completed]
[09:24:07] Running tasks for **/*.+(js|json|less|css|ts) [completed]
25h25h
husky > pre-commit hook failed (add --no-verify to bypass)
```

<details>
<summary>verbose npm output</summary>


```
husky > npm run -s precommit (node v8.5.0)


> kcd-scripts@0.0.0-semantically-released precommit E:\Projects\repos\kcd-scripts
> node src precommit

25l[09:25:15] Running tasks for **/*.+(js|json|less|css|ts) [started]
[09:25:15] Running tasks for .all-contributorsrc [started]
[09:25:15] Running tasks for .all-contributorsrc [skipped]
[09:25:15] → No staged files match .all-contributorsrc
[09:25:15] kcd-scripts format [started]
[09:25:16] kcd-scripts format [completed]
[09:25:16] kcd-scripts lint [started]
[09:25:18] kcd-scripts lint [completed]
[09:25:18] kcd-scripts test --findRelatedTests [started]
[09:25:20] kcd-scripts test --findRelatedTests [completed]
[09:25:20] git add [started]
[09:25:20] git add [completed]
[09:25:20] Running tasks for **/*.+(js|json|less|css|ts) [completed]
25h25hnpm WARN invalid config access=undefined
npm WARN invalid config also=undefined
npm WARN invalid config https-proxy=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN invalid config init-author-url=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN invalid config only=undefined
npm WARN invalid config proxy=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN onload-script failed to require onload script undefined
npm WARN onload-script Error: Cannot find module 'undefined'
npm WARN onload-script     at Function.Module._resolveFilename (module.js:527:15)
npm WARN onload-script     at Function.Module._load (module.js:476:23)
npm WARN onload-script     at Module.require (module.js:568:17)
npm WARN onload-script     at require (internal/module.js:11:18)
npm WARN onload-script     at cb (C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:225:11)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:356:16
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:81:7
npm WARN onload-script     at Array.forEach (<anonymous>)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:80:13
npm WARN onload-script     at f (C:\nvm\v8.5.0\node_modules\npm\node_modules\once\once.js:25:25)
npm WARN onload-script  { Error: Cannot find module 'undefined'
npm WARN onload-script     at Function.Module._resolveFilename (module.js:527:15)
npm WARN onload-script     at Function.Module._load (module.js:476:23)
npm WARN onload-script     at Module.require (module.js:568:17)
npm WARN onload-script     at require (internal/module.js:11:18)
npm WARN onload-script     at cb (C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:225:11)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:356:16
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:81:7
npm WARN onload-script     at Array.forEach (<anonymous>)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:80:13
npm WARN onload-script     at f (C:\nvm\v8.5.0\node_modules\npm\node_modules\once\once.js:25:25)
npm WARN onload-script   stack: 'Error: Cannot find module \'undefined\'\n    at Function.Module._resolveFilename (module.js:527:15)\n    at Function.Module._load (module.js:476:23)\n    at Module.require (module.js:568:17)\n    at require (internal/module.js:11:18)\n    at cb (C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\npm.js:225:11)\n
    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\npm.js:356:16\n    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\config\\core.js:81:7\n    at Array.forEach (<anonymous>)\n    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\config\\core.js:80:13\n    at f (C:\\nvm\\v8.5.0\\node_modules\\npm\\node_modules\\once\\once.js:25:25)',
npm WARN onload-script   code: 'MODULE_NOT_FOUND' }
npm WARN invalid config access=undefined
npm WARN invalid config also=undefined
npm WARN invalid config https-proxy=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN invalid config init-author-url=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN invalid config only=undefined
npm WARN invalid config proxy=undefined
npm WARN invalid config Must be a full url with 'http://'
npm WARN onload-script failed to require onload script undefined
npm WARN onload-script Error: Cannot find module 'undefined'
npm WARN onload-script     at Function.Module._resolveFilename (module.js:527:15)
npm WARN onload-script     at Function.Module._load (module.js:476:23)
npm WARN onload-script     at Module.require (module.js:568:17)
npm WARN onload-script     at require (internal/module.js:11:18)
npm WARN onload-script     at cb (C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:225:11)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:356:16
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:81:7
npm WARN onload-script     at Array.forEach (<anonymous>)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:80:13
npm WARN onload-script     at f (C:\nvm\v8.5.0\node_modules\npm\node_modules\once\once.js:25:25)
npm WARN onload-script  { Error: Cannot find module 'undefined'
npm WARN onload-script     at Function.Module._resolveFilename (module.js:527:15)
npm WARN onload-script     at Function.Module._load (module.js:476:23)
npm WARN onload-script     at Module.require (module.js:568:17)
npm WARN onload-script     at require (internal/module.js:11:18)
npm WARN onload-script     at cb (C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:225:11)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\npm.js:356:16
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:81:7
npm WARN onload-script     at Array.forEach (<anonymous>)
npm WARN onload-script     at C:\nvm\v8.5.0\node_modules\npm\lib\config\core.js:80:13
npm WARN onload-script     at f (C:\nvm\v8.5.0\node_modules\npm\node_modules\once\once.js:25:25)
npm WARN onload-script   stack: 'Error: Cannot find module \'undefined\'\n    at Function.Module._resolveFilename (module.js:527:15)\n    at Function.Module._load (module.js:476:23)\n    at Module.require (module.js:568:17)\n    at require (internal/module.js:11:18)\n    at cb (C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\npm.js:225:11)\n
    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\npm.js:356:16\n    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\config\\core.js:81:7\n    at Array.forEach (<anonymous>)\n    at C:\\nvm\\v8.5.0\\node_modules\\npm\\lib\\config\\core.js:80:13\n    at f (C:\\nvm\\v8.5.0\\node_modules\\npm\\node_modules\\once\\once.js:25:25)',
npm WARN onload-script   code: 'MODULE_NOT_FOUND' }

> kcd-scripts@0.0.0-semantically-released validate E:\Projects\repos\kcd-scripts
> node src validate

npm ERR! file undefined
npm ERR! path undefined
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn undefined
npm ERR! kcd-scripts@0.0.0-semantically-released validate: `node src validate`
npm ERR! spawn undefined ENOENT
npm ERR!
npm ERR! Failed at the kcd-scripts@0.0.0-semantically-released validate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! code ELIFECYCLE
npm ERR! errno -4058
npm ERR! kcd-scripts@0.0.0-semantically-released validate: `node src validate`
npm ERR! Exit status -4058
npm ERR!
npm ERR! Failed at the kcd-scripts@0.0.0-semantically-released validate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\suhas.karanth\AppData\Roaming\npm-cache\_logs\2017-09-19T03_55_21_615Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! kcd-scripts@0.0.0-semantically-released precommit: `node src precommit`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the kcd-scripts@0.0.0-semantically-released precommit script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\suhas.karanth\AppData\Roaming\npm-cache\_logs\2017-09-19T03_55_21_791Z-debug.log

husky > pre-commit hook failed (add --no-verify to bypass)
```

</details>

<br>

Problem description:

This is related to the issue I had faced with `opt-cli` - https://github.com/ta2edchimp/opt-cli/pull/213.
Invoking `spawn` with env having `undefined` values breaks `npm`.

Suggested solution:

While I know how to fix the issue, in `kcd-scripts`, I am wondering if the fix should be in `cross-spawn` or even `npm`.

--- a/.all-contributorsrc
+++ b/.all-contributorsrc
@@ -25,7 +25,8 @@
       "avatar_url": "https://avatars2.githubusercontent.com/u/22251956?v=4",
       "profile": "https://github.com/sudo-suhas",
       "contributions": [
-        "code"
+        "code",
+        "bug"
       ]
     }
   ]

--- a/README.md
+++ b/README.md
@@ -107,7 +107,7 @@ here! Again, this is a very specific-to-me solution.
 Thanks goes to these people ([emoji key][emojis]):
 
 <!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
-| [<img src="https://avatars.githubusercontent.com/u/1500684?v=3" width="100px;"/><br /><sub>Kent C. Dodds</sub>](https://kentcdodds.com)<br />[💻](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Code") [📖](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Documentation") [🚇](#infra-kentcdodds "Infrastructure (Hosting, Build-Tools, etc)") [⚠️](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Tests") | [<img src="https://avatars2.githubusercontent.com/u/22251956?v=4" width="100px;"/><br /><sub>Suhas Karanth</sub>](https://github.com/sudo-suhas)<br />[💻](https://github.com/kentcdodds/kcd-scripts/commits?author=sudo-suhas "Code") |
+| [<img src="https://avatars.githubusercontent.com/u/1500684?v=3" width="100px;"/><br /><sub>Kent C. Dodds</sub>](https://kentcdodds.com)<br />[💻](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Code") [📖](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Documentation") [🚇](#infra-kentcdodds "Infrastructure (Hosting, Build-Tools, etc)") [⚠️](https://github.com/kentcdodds/kcd-scripts/commits?author=kentcdodds "Tests") | [<img src="https://avatars2.githubusercontent.com/u/22251956?v=4" width="100px;"/><br /><sub>Suhas Karanth</sub>](https://github.com/sudo-suhas)<br />[💻](https://github.com/kentcdodds/kcd-scripts/commits?author=sudo-suhas "Code") [🐛](https://github.com/kentcdodds/kcd-scripts/issues?q=author%3Asudo-suhas "Bug reports") |
 | :---: | :---: |
 <!-- ALL-CONTRIBUTORS-LIST:END -->
 

--- a/src/index.js
+++ b/src/index.js
@@ -19,6 +19,8 @@ if (!scriptPath) {
   console.log('Perhaps you need to update kcd-scripts?')
 }
 
+// this is required to address an issue in cross-spawn
+// https://github.com/kentcdodds/kcd-scripts/issues/4
 const env = Object.keys(process.env)
   .filter(key => process.env[key] !== undefined)
   .reduce(

--- a/src/index.js
+++ b/src/index.js
@@ -19,11 +19,21 @@ if (!scriptPath) {
   console.log('Perhaps you need to update kcd-scripts?')
 }
 
+const env = Object.keys(process.env)
+  .filter(key => process.env[key] !== undefined)
+  .reduce(
+    (envCopy, key) => {
+      envCopy[key] = process.env[key]
+      return envCopy
+    },
+    {
+      [`SCRIPTS_${script.toUpperCase()}`]: true,
+    },
+  )
+
 const result = spawn.sync(executor, [scriptPath, ...args], {
   stdio: 'inherit',
-  env: Object.assign({}, process.env, {
-    [`SCRIPTS_${script.toUpperCase()}`]: true,
-  }),
+  env,
 })
 
 if (result.signal) {

